Simulations based on age/mass prior model

karllark opened the pull request:
The number of stars in a simulation can now be computed from age/mass priors.  This is needed to get simulations that work for testing the MegaBEAST (and MATCH).  It is still possible to set the number of stars to simulate explicitly (so - two options now).

To provide the age/mass priors to the simulation code, a new file was created to save information on the BEAST model/run.  This is the "beast_info" file and is saved as as ASDF file as this provides a *very* easy way to store dictionaries.  This beast info file will be used by the MegaBEAST as well.

As part of this work, it was noticed that the plot created from `plot_indiv_plot` looked a bit odd - specifically for the lines/area for the stellar+dust+bias fluxes.  This was traced to the use of symlogs to handle the large dynamic range and negative values for these fluxes.  The symlog threshold (transition between log and linear regimes) was set much too high.  A threshold of 1e-20 was found to work much better.  A symlog module was added to provide the needed functions.  The flux plot in `plot_indiv_fit` was changed to be a symlog plot (supported by matplotlib!) to show the positive and negative fluxes (model and observed).  The default max number of bins in the 1D PDFs has been changed from 100 to 200 to better resolve the 1D PDFs - especially those based on symlogs.

Docs updated and tests included.

Closes #593.

kmcquinn commented:
**Review of PR by McQuinn:**
1) The output file gives the simulated band_vega as vega magnitudes with zero and negative fluxes given as -99.999. MATCH, however, uses 99.999 as the non-recovered mag values. I think having the minus sign will cause problems when fitting BEAST simulated photometry as a test of the SFH recovery process. So, if it doesn't cause issues with BEAST, it would be preferable to have the vega band mag with zero and negative fluxes listed as 99.999  (i.e., no minus sign). 

2) There are some leftover formatting/typos from the previous documentation:
This provides

the star formation rate verus time in the age prior combined with the

is given by the --nsim parameter.

3) Is there a sample beastinfo ASDF file that can be added to the documentation? What information does this file contain?

4) One option in the simulations is which filter to use for the completeness function:
" The filter to use for the completeness function is given by the --compl_filter parameter (default=F475W). Set compl_filter=max to use the max completeness value across all the filters. "

How does this impact the simulated photometry? Any guidance on how to choose which filter to use? Or is this similar to running the BEAST whereh F475W is just generally used?

Thanks!

lea-hagen commented on beast/physicsmodel/prior_weights_stars.py:
```diff
@@ -83,7 +87,8 @@ def compute_age_prior_weights(logages, age_prior_model):
         )
 
     # normalize to avoid numerical issues (too small or too large)
-    age_weights /= np.average(age_weights)
+    # do not normalize as the absolute level now matters for simulations and megaBEAST
+    # age_weights /= np.average(age_weights)
```
Since normalization was originally to avoid numerical issues, do we need to worry about that now?

karllark replied:
I just did a blanket normalization to all the priors in multiple places.  I checked and it should not be an issue for any reasonable age prior.

lea-hagen commented on beast/plotting/plot_toothpick_details.py:
```diff
@@ -60,6 +62,10 @@ def plot_toothpick_details(asts_filename, seds_filename, savefig=False):
 
         # not all bins are filled with good data
         ngbins = model._nasts[i]
+        print(model._minmax_asts[:, i])
+
+        print(model._fluxes[0:ngbins, i])
```
Do these still need to be here?

karllark replied:
Removed.

lea-hagen reviewed (changes requested):
Looking good!  I have a few questions/comments.
* As best I can tell, the purpose of `beast_info` is to keep a record of the beast settings, so that later it's easy to find the settings that were used.  Is that correct?  If so, what is the advantage of transferring them to an asdf file rather than, say, running `shutil.copy2(beast_settings_file, project_dir)` and then using the existing beast_settings infrastructure?
* As `gen_SimObs_from_sedgrid` keeps getting more substantial (I think in the past it was maybe 20-30 lines?), I'm wondering if it should move to its own file.  That way `observations.py` doesn't get too unwieldy.
* Could you show an example of the new `plot_indiv_fit`?

lea-hagen commented on docs/simulations.rst:
```diff
@@ -2,27 +2,41 @@
+(mass prior).  This number of total stars formed is then the total mass divided
+by the average mass computed from the mass prior model.  This total number of
+stars is simulated and then the completeness function is applied to remove
+all the simulated stars that would not be observed.  This provides
+
+the star formation rate verus time in the age prior combined with the
+
+is given by the `--nsim` parameter.
```
Woops, forgot to add that this has a few extra line breaks!

lea-hagen replied:
(which I also just realized was in @kmcquinn's review, too)

karllark replied:
Fixed.

karllark commented:
Thanks for the review @kmcquinn.  Responses:

1) Done.

2) Fixed.

3) This file contains saved meta info from the run to better log what information that is needed for the simulations *and* by the MegaBEAST.  I expect the content of this file will expand significantly to as the MegaBEAST is fully developed.  Definitely should say something in the documentation.  I've added a line to the `outputs` docs.

4) I am not sure of the right answer here.  I'm also not sure that it makes a huge difference.  The reason is that the toothpick model is an approximation of what is really happening when the photometry is measured.  In reality, there is only one completeness value for a particular model star that depends on the flux of that model in *all* the bands (i.e., brightness and shape).  We aren't doing enough ASTs currently to define the completeness in that space.  Hence -> approximation using toothpick model.  My gut says that `max` is probably the right thing to use for the completeness, but I don't have any quantitative proof of this yet.

karllark commented:
> Looking good! I have a few questions/comments.
> 
> * As best I can tell, the purpose of `beast_info` is to keep a record of the beast settings, so that later it's easy to find the settings that were used.  Is that correct?  If so, what is the advantage of transferring them to an asdf file rather than, say, running `shutil.copy2(beast_settings_file, project_dir)` and then using the existing beast_settings infrastructure?

I did think about this.  But I thought that a separate output info file would be better as then we can put anything we want in it, not just what was passed in via the settings.  I already have done this by including not just the prior models, but also the 1D vectors of the actual grid values for the dust parameters. 

> * As `gen_SimObs_from_sedgrid` keeps getting more substantial (I think in the past it was maybe 20-30 lines?), I'm wondering if it should move to its own file.  That way `observations.py` doesn't get too unwieldy.

Could be done.  I guess my idea was that these are simulations of the observations so they belonged there.  We have lots of longer files, so I just left it there.  We should think about this more globally for all the BEAST files.  Moving things can break other scripts/code, so...

> * Could you show an example of the new `plot_indiv_fit`?

Yeah, good point.  I've changed the example to be from the metal_small example instead of phat_small.  phat_small is something we just shouldn't advertise anymore anyway.

karllark commented on beast/tools/symlog.py:
```diff
@@ -0,0 +1,48 @@
+    y_symlog : float
+        symlog of y
+    """
+    return np.sign(y) * np.log10(1.0 + np.abs(y / C))
```
I'm not sure this is an issue anymore given that C is 1e-20.  We shouldn't see anything really small anymore.  This should be a separate issue so that the math can be checked.

lea-hagen replied:
Fair point.  I just did a quick test with y=1e-30, and the difference in the two methods is 0.0000008%, so I agree that it's fine!

karllark commented on beast/tools/symlog.py:
```diff
@@ -0,0 +1,48 @@
+    y_linear : float
+        linear version of y
+    """
+    return np.sign(y) * C * (-1.0 + 10.0 ** np.abs(y))
```
Same reply as above.

karllark commented on beast/tools/tests/test_symlog.py:
```diff
@@ -0,0 +1,26 @@
+    # test going to symlog
+    np.allclose(symlog(y), exp_symlog)
+
+    # test converting back to liner
```
Fixed.

karllark commented:
@lea-hagen & @kmcquinn : ok to merge this PR?  I.e., were my responses to your reviews sufficient?

kmcquinn commented:
@karllark Just want to clarify something about the beastinfo ASDF file.  Is this something the user creates? Or is this generated by BEAST based on the inputs to the simulation? If it's the latter, then I'm happy with the responses. Thank you!

karllark commented:
> @karllark Just want to clarify something about the beastinfo ASDF file. Is this something the user creates? Or is this generated by BEAST based on the inputs to the simulation? If it's the latter, then I'm happy with the responses. Thank you!

It is generated by the BEAST.  It is way for the BEAST to save info that is useful later.  Definitely something that a users should *not* create by hand.

lea-hagen reviewed (approved):
Changes look great!

kmcquinn commented:
I think it looks good, too! I'm excited to try it out!
